internal/uvm/scsi/manager.go

```diff
@@ -0,0 +1,322 @@
+package scsi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/protocol/guestresource"
+	"github.com/Microsoft/hcsshim/internal/security"
+	"github.com/Microsoft/hcsshim/internal/verity"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrNoAvailableLocation indicates that a new SCSI attachment failed because
+	// no new slots were available.
+	ErrNoAvailableLocation = errors.New("no available location")
+	// ErrNotInitialized is returned when a method is invoked on a nil [Manager].
+	ErrNotInitialized = errors.New("SCSI manager not initialized")
+	// ErrAlreadyReleased is returned when [Mount.Release] is called on a Mount
+	// that had already been released.
+	ErrAlreadyReleased = errors.New("mount was already released")
+)
+
+// Manager is the primary entrypoint for managing SCSI devices on a VM.
+// It tracks the state of what devices have been attached to the VM, and
+// mounted inside the guest OS.
+type Manager struct {
+	attachManager *attachManager
+	mountManager  *mountManager
+}
+
+// Slot represents a single SCSI slot, consisting of a controller and LUN.
+type Slot struct {
+	Controller uint
+	LUN        uint
+}
+
+// NewManager creates a new Manager using the provided backend [Attacher],
+// [Mounter], and [Unplugger], as well as other configuration parameters.
+//
+// guestMountFmt is the format string to use for mounts of SCSI devices in
+// the guest OS. It should have a single %d format parameter.
+//
+// reservedSlots indicates which SCSI slots to treat as already used. They
+// will not be handed out again by the Manager.
+func NewManager(
+	attacher Attacher,
+	mounter Mounter,
+	unplugger Unplugger,
+	numControllers int,
+	numLUNsPerController int,
+	guestMountFmt string,
+	reservedSlots []Slot,
+) *Manager {
+	am := newAttachManager(attacher, unplugger, numControllers, numLUNsPerController, reservedSlots)
+	mm := newMountManager(mounter, guestMountFmt)
+	return &Manager{am, mm}
+}
+
+// MountConfig specifies the options to apply for mounting a SCSI device in
+// the guest OS.
+type MountConfig struct {
```

Comment:
I originally had a `ReadOnly` field on `MountConfig`, but thought it made things more verbose with (potentially) little benefit. We already have the `readOnly` parameter on the exporter `Manager` calls, and I couldn't think of a reason we would want the attachment and mount to disagree on the read-only-ness.

That said, I'd be fine with adding back `ReadOnly` on `MountConfig`.

Reply:
I see, I was thinking that we'd remove the 'readonly' param to the Manager calls. If we decide to continue using the MountConfig to indicate if an attached device should also be mounted, then I think we'd want to keep this field off of the MountConfig and retain it in the Manager calls. Otherwise, maybe we can revisit this?

Reply:
I wouldn't want the attachment read-only state to be determined by a field on the MountConfig, though. That would be confusing since then a "mount-specific" field would be determining "attachment-specific" state.

But we could go to having read-only both as an argument on the method (to determine attachment state) and having it on MountConfig. Or we could get rid of the `Add*Disk` methods entirely, and just expose `add` directly to clients.

Reply:
hmm this may be in part from my own confusion about the MountConfig. I think I was confused in thinking that if we no longer used MountConfig to indicate whether to mount the attached device or not, then we'd move to using it in a more generic way as the configuration for the whole process of attaching and mounting, rather than just for the mounting. Which I don't think is what we'd want to do. 

As an idea, what do you think about also exposing an "AttachConfig" that mirrors that of the internal attach config? In the future if we're exposing more about an attachment vs a mount, we might want this config struct anyways. What do you think? We could even (now or in a follow up later) combine the different `Add*Disk` methods by exposing "PassThru", "VirtualDisk", "ExtensibleVirtualDisk" as options on the "AttachConfig" and handling them in a single `AddSCSI` call.

Reply:
I pushed a couple new experimental commits to try out something like this. It's certainly more explicit, but my concern is it's also more verbose, and potentially not as ergonomic. Let me know what you think.

### Commit [scsi: Remove VMAccessType, move to clients](https://github.com/microsoft/hcsshim/pull/1744/commits/276173af18fe31d573c4482431924c4f079339f0)

Removes the VMAccessType handling from the scsi package, and instead relies on the client to handle it. It didn't really make sense for the scsi package to handle this anyway, and it lets us unify how it handles different attachment types better.

### Commit [try new SCSIManager exports](https://github.com/microsoft/hcsshim/pull/1744/commits/6d799cba27379030fe186e347f159cc0321de406)

Tests a a new approach where we just expose `(*scsi.Manager).Add` which takes a `*scsi.AttachConfig` and `*scsi.MountConfig`. This moves more responsibility to the guest to set things up like setting the attachment type, and determing verity info.

Reply:
What are your concerns around it potentially being less ergonomic? Was there another way you think we might want to expose the attachment vs mount calls that would make that better?

It is more verbose, but it's not requiring the caller to know any more information about how the attachment and mount will be setup than it already needed to know really, so I'm not sure I see why it would be less ergonomic.

Reply:
I think verbosity can contribute negatively to ergonomics. It can make it harder to at a glance tell what we're trying to do. Needing to pass a read-only parameter in two different places (but also you probably always want them to match) is also awkward.

Generally the goal of the `SCSIManager` was to provide a friendlier abstraction on top of the `attachManager` and `mountManager`. With the code above, we're not really getting much abstraction.

I'm not exactly sure where I would want this interface to go long-term, but a few thoughts:
- Decouple mounts from attachments, potentially with a single mount manager that can mount any type of storage (SCSI, VPMEM, etc.). It just needs a way to determine the guest path for that storage (at least on LCOW) so it can mount it.
- Maybe add some sort of high-level `StorageManager` that provides a nice interface on top of attaching/mounting all storage.

Reply:
Also interested in getting more feedback on this. @helsaawy any thoughts?

Reply:
Based on feedback, I have largely moved back to the original version of the `scsi.Manager` interface. I did notice we never utilize "group access" to an attachment, so I removed that capability. There is now simply a `vmID` parameter to `AddVirtualDisk`/`AddPhysicalDisk` which, if set, causes that VM to be granted access to the disk.

This current implementation is not perfect, but I believe it is a large improvement over the previous sprawl of SCSI in terms of maintainability, and, importantly, it delivers us the ability to mount multiple partitions of an attachment. We can continue to iterate on and improve our storage management in the future, drawing on ideas from discussion here, as well as what I've put in the README. I think we will also likely learn more about how we want this code to evolve as we work on decoupling UVM/container management, and that will help inform future work. :)

I'm hoping we can get a consensus on taking this current implementation for now, without any major structural changes.

Reply:
This would break the current semantic of passing a `nil` `MountConfig` to mean "attach but do not mount in the guest". I can look at if there is another way to represent that. We can also iterate on this interface in the future if we have more ideas later.

Reply:
The issue I'm running into here is that if we make a:

```golang
type Config struct {
    ReadOnly bool
    Encrypted bool
    Options []string
}
```

Then there is no way to do attach-without-mount.

If we instead make it:

```golang
type Config struct {
    Mount bool
    ReadOnly bool
    Encrypted bool
    Options []string
}
```

Then I feel the interface is a bit awkward, and also error-prone because you could set `Encrypted` and `Options`, but they won't do anything unless `Mount` is `true`.

So then we could do:

```golang
type Config struct {
    ReadOnly bool
    Mount *MountConfig
}

type MountConfig struct {
    Encrypted bool
    Options []string
}
```

But this seems like it has the same concern you originally didn't like, namely that there is both a `ReadOnly` value, and a separate `MountConfig`.

So I'm not sure how to handle this, and somwhat inclined to keep it as is for now, unless we can find an approach that is clearly better.

Reply:
but at the very least, this approach would move the `ReadOnly` from a function parameter to a config field, so it would be consistent.

The other approach would be to keep the full `Config` struct, but make it private and instead use `type Option func(*config) error` options, so users would only have `ReadOnly`, `Mount(encrypted bool, mountOptions []string)`, options, and couldn't come up with invalid config combos

Reply:
For the sake of getting this in without further churn I am planning to merge as-is. However, I think the functional options approach sounds reasonable here, though may require a bit of reworking of the manager functions to do nicely. I'm definitely open to this change being made later.